src/nsrt_learning/side_predicate_learning.py

```diff
@@ -394,59 +400,92 @@ def _sideline(self) -> List[PartialNSRTAndDatastore]:
-                    assert new_pnad is not None, "1:1 assumption violated"
-                    assert new_pnad.option_spec == pnad.option_spec
+                substitution_exists = False
+                # We start by checking if any of the PNADs associated with the
+                # demonstrated option are able to match this transition.
+                for pnad in pnads_for_option:
+                    ground_op = self._find_unification(necessary_add_effects,
+                                                       pnad, segment)
+                    if ground_op is not None:
+                        obj_to_var = dict(
+                            zip(ground_op.objects, pnad.op.parameters))
+                        if param_opt_to_nec_pnad.get(option.parent) is None:
+                            param_opt_to_nec_pnad[option.parent] = [pnad]
+                        substitution_exists = True
+                        break
+
+                # If we weren't able to find a substitution, we need to try
+                # specifize each of our PNADs.
+                if not substitution_exists:
+                    should_split_off_new_pnad = True
+                    for pnad in pnads_for_option:
+                        # We now try to make the PNAD's add effects more
+                        # specific. Since we have started with the most
+                        # general operators, and we are making them more
+                        # specific, it is guaranteed that such a
+                        # grounding exists, otherwise our 1:1
+                        # option:operator assumption is violated.
+                        new_pnad = self._try_specializing_pnad(
+                            necessary_add_effects, pnad, segment)
+                        if new_pnad is not None:
+                            assert new_pnad.option_spec == pnad.option_spec
+                            if param_opt_to_nec_pnad.get(
+                                    option.parent) is not None:
+                                param_opt_to_nec_pnad[option.parent].remove(
+                                    pnad)
+                            del pnad
+                            should_split_off_new_pnad = False
+                            break
+
+                    # If we were unable to specifize any of the PNADs, we need
+                    # to split from the most general PNAD and make a new PNAD
+                    # to cover these necessary add effects.
+                    if should_split_off_new_pnad:
+                        new_pnad = self._try_specializing_pnad(
+                            necessary_add_effects,
+                            param_opt_to_general_pnad[option.parent], segment)
+                        assert new_pnad is not None
+
+                    # Add the new PNAD to the dictionary mapping options to
+                    # PNADs.
+                    assert isinstance(new_pnad, PartialNSRTAndDatastore)
```

Comment:
do we need this assert? can we remove it?

Reply:
necessary for mypy to pass

Reply:
change to `assert new_pnad is not None`? does mypy still pass?

Reply:
huh - turns out this works (i thought i tried this before and it didn't work)